lib/internal/modules/esm/translators.js

```diff
@@ -152,18 +151,9 @@ translators.set('module', async function moduleStrategy(url) {
   return module;
 });
 
-function enrichCJSError(err) {
-  if (err == null || ObjectGetPrototypeOf(err) !== SyntaxErrorPrototype) {
-    return;
-  }
-  const stack = StringPrototypeSplit(err.stack, '\n');
-  /*
-  * The regular expression below targets the most common import statement
-  * usage. However, some cases are not matching, cases like import statement
-  * after a comment block and/or after a variable definition.
-  */
-  if (StringPrototypeStartsWith(err.message, 'Unexpected token \'export\'') ||
-    RegExpPrototypeTest(/^\s*import(?=[ {'"*])\s*(?![ (])/, stack[1])) {
+function enrichCJSError(err, content) {
+  if (err != null && ObjectGetPrototypeOf(err) === SyntaxErrorPrototype &&
+      hasEsmSyntax(content)) {
```

Comment:
We could avoid an unnecessary readfile on 193 here on the non-syntax error cases by making the signature `enrichCJSError(err, content, filename)` permitting `content` as undefined corresponding to `filename` being set. Then move the `readFile` into this line instead.

Reply:
Instead of moving the source of truth, could hasEsmSyntax memoize?

Reply:
It will likely only be called once per source as it's only called on erroring currently.

Reply:
Then where’s the extra readFile coming from?

Reply:
We don't keep executed CJS sources in memory, hence it requires a re-read. The OS-level FS caching should ensure this remains fast for the most part though. I was just suggesting a path to avoid unnecessary re-reads for eg normal execution errors.